### PR TITLE
Use integer as kernel_quantizer to mark layers that should use Bop

### LINKED ISSUE #306: RFC: Quantized Variables
This Request for Comments (RFC) outlines a proposed addition to the API of `larq`. It would be good to get feedback on this, so please ask questions if I need to elaborate more in some areas. WIP implementation can be viewed at #312.

## Objective

`larq` has a concept of quantized layers that define a [quantized operation using quantized inputs and variables](https://larq.dev/guides/key-concepts/#quantized-layers). Quantized layers are currently the only official way to interact with `larq`. E.g. accessing the precision of quantized variables or inputs are cumbersome to get and require checking `layer.quantized_latent_weights` and `layer.quantizers`. This often requires downstream code that needs access to variables to rely on private methods. We can fix this by either introducing convenience functions and a stable API of quantized layers to access quantized variables or extend the API with the concept of a quantized variable. This document will elaborate on the latter since it allows new use cases for `larq` and gives users more fine grained control when using custom layers or no layers at all. This concept can be extended to `tf.Tensor` for quantized activations which we can discuss in a future RFC.

## Design Proposal

There are a few ways we could implement this functionality, for the initial proposal I closely followed the implementation of [`AutoCastVariable`](https://github.com/tensorflow/tensorflow/blob/master/tensorflow/python/keras/mixed_precision/experimental/autocast_variable.py#L28) which only wraps `tf.Variable`.
While the variable acts like a quantized variable the TensorFlow data type will still be floating-point in order to be compatible with all TensorFlow operations. This is typically referred to as fake quantization.

```python
class QuantizedVariable(tf.Variable):
    def __init__(self, variable, quantizer=None, precision=None):
        """Creates an QuantizedVariable instance.

        # Arguments
        variable: A floating-point resource variable to wrap.
        quantizer: An optional quantizer to transform the floating-point
            variable to a fake quantized variable.
        precision: An optional integer defining the precision of the quantized
            variable. If `None`, `quantizer.precision` is used.
        """
        pass

    @property
    def precision(self):
        """Returns the integer precision of the quantized variable."""
        pass

    def latent_value(self):
        """Returns the latent value of the variable."""
        pass

    # Overload all operators to use the quantized version and delegate all assign
    # operations to the latent weight.
```

This variable would act like a quantized variable in the forward pass by overloading the default operators like `__add__`, `__mul__`, etc. but would delegate updates via `self.assign*` to the underlying latent variable. This would make reasoning about quantized variables a lot easier and would allow features like the model converter or model summary to have easy access to all information.

### How can we handle variable saving?

For the purpose of resuming training the latent variable should be saved to the checkpoint. For checkpoints we can set this behaviour by delegating `_gather_saveables_for_checkpoint` to the wrapped variable. Will this also work for Keras when saving to `.h5` which might call `get_weights` during saving?

### How should we handle true (fake) quantized variables like in Bop?

The most straight forward approach would be to keep the current design and use `quantizer=None` and `precision=1`, though an alternative would be to introduce an extra class.

### How can we support quantizers with trainable variables?

We currently don't have  a quantizer with trainable variables in `larq` core, but probably add one in the near future (e.g. 8-bit). It would be good to see how this would work with this approach. See also #89 

### Don't quantize the value of the variable by default

This would have the advantage that we won't run into any problems with variable saving or other side effect we might neglect. This would make usability a bit worse since we would need a context manager to define a scope within which the variable would return the quantized value.

### Should we wrap an existing `tf.Variable` or create a new one?

I think the approach of wrapping an existing `tf.Variable` is easier to implement and maintain. To me, it also doesn't have major downsides from an user perspective.

### Extend quantized layers with methods to return quantized versions of the variables

We could extend the `larq` layers to allow better access to the underlying variables. However I think ultimately this would be a worse abstraction than introducing a new variable type.

## Testing process

Since this touches a few Keras and TensorFlow internals we need very good testing for it. Below I highlighted some code paths that might have non obvious interactions with the `QuantizedVariable` and need extra care:

- Multi GPU training and `tf.DistributedVariable`
- Keras Float 16 training using `AutoCastVariable`
- TPUStrategy (will only be supported by TensorFlow 2.1+, so no current concern)
- Model checkpointing and saving (see above)
- Gradient computation in eager, functional and legacy graph mode

### PR DESCRIPTION
**Pull request by lgeiger**
As discussed in https://github.com/larq/larq/issues/306#issuecomment-553978646, this allows to use integers as `kernel_quantizers` in layers to mark them to be used by external optimizers like Bop.

This will be a breaking change because previously `Bop` was applied to **all** quantized layers which made it impossible to use `larq` layers with different precisions in a network. Now `kernel_quantizers=1` needs to be set on all layers that should be optimized by `Bop`. `kernel_quantizers="ste_sign"` keeps working too, but will raise a warning message.

Please review carefully since this is a user level API change.
Do you think it would useful to add a `Bop.legacy_is_binary_variable` method that exposes the deprecated old behaviour to ease the migration?

Closes #306

### CODE REVIEW

File: larq/optimizers.py [hunk 1]
```diff
@@ -205,6 +206,10 @@ class Bop(tf.keras.optimizers.Optimizer):
     Bop is a latent-free optimizer for Binarized Neural Networks (BNNs) and
     Binary Weight Networks (BWN).
 
+    To use Bop to train a model set `kernel_quantizer=1` for the layers where
+    binarization should be handled by Bop or use custom predicate function in the
+    `CaseOptimizer`.
```
Comment (leonoverweel) [inline larq/optimizers.py]: Suggestion:

To use Bop to train a model, set `kernel_quantizer=1` for the layers where training and binarization should be handled by Bop. This is the easiest, recommended approach. Alternatively, you can define a custom predicate function for the [`CaseOptimizer`](https://larq.dev/api/optimizers/#caseoptimizer) that captures the set of layers you want to assign to Bop (see Bop's [`is_binary_var()`](https://github.com/larq/larq/blob/master/larq/optimizers.py#L288) for example.)

File: larq/optimizers.py [hunk 2]
```diff
@@ -286,11 +292,17 @@ def from_config(cls, config, custom_objects=None):
...
 
         # Arguments
         var: a `tf.Variable`.
         """
-        return "/kernel" in var.name and "quant_" in var.name
+        precision = getattr(var, "precision", None)
+        if precision == 1 and getattr(var, "quantizer", None) is not None:
+            warnings.warn(
+                f"Calling Bop with `quantizer={var.quantizer}`; this might result in "
+                "unexpected behaviour or won't do anything."
```
Comment (leonoverweel) [inline larq/optimizers.py]: Suggested addition:

To assign a layer to Bop, set `kernel_quantizer=1`.

File: larq/quantizers_test.py [hunk 3]
```diff
@@ -56,7 +56,7 @@ def test_serialization_cls(self, ref_fn):
 
     def test_invalid_usage(self):
         with pytest.raises(ValueError):
-            lq.quantizers.get(42)
+            lq.quantizers.get(42.0)
```
Comment (leonoverweel) [inline larq/quantizers_test.py]: Hehe

File: larq/layers.py [hunk 4]
```diff
@@ -67,7 +67,9 @@ class QuantDense(QuantizerBase, tf.keras.layers.Dense):
         no activation is applied (`a(x) = x`).
     use_bias: Boolean, whether the layer uses a bias vector.
     input_quantizer: Quantization function applied to the input of the layer.
-    kernel_quantizer: Quantization function applied to the `kernel` weights matrix.
+    kernel_quantizer: Quantization function applied to the `kernel` weights matrix
+        or integer defining the desired precision if quantization should be handled
+        by an external optimizer.
```
Comment (koenhelwegen) [inline larq/layers.py]: I would use "specialized optimizer" instead of "external optimizer"

**Comment by lgeiger** [comment]
>  Such a user may simply set `kernel_quantizer=1`, throw in Adam, and expect it to result in binary weights, which if I understand correctly now would not be the case.

That's correct. I agree that the option of passing a desired precision as quantizer isn't the best from an API perspective. Though in your example I am afraid that we cannot avoid this problem in all cases (i.e. `kernel_quantizer=None` or  `kernel_quantizer=lambda x: 2*x` also don't result in quantized weights).

> What I would propose as a solution is to modify the behaviour of our QuantizedVariable class such that if the precision is set to 1 (or any other integer), the variable will always be quantized on the forward pass. As default quantizer I would propose the STE (or stepwise functions between -1 and 1 with clipped identity as backward pass for higher precisions).

I am against this solution due to two main issues: 1) Since this would introduces unnecessary operations in the graph it might degrade training performance. 2) Implicitly quantizing might not be what a users is trying to achieve and is also something that is not clearly visible via the API. E.g. You could think of a scaling "quantizer" `lambda x: alpha*x` with a trainable variable `alpha` where `x` is updated via Bop and `alpha` via a normal optimizer or a moving average (sorry for the contrived example). Things like that would not be possible to implement since `larq` would still quantize it to -1, +1.

How would you feel about a `NoOpQuantizer` (probably needs a better name), that is used instead of an integer? It could to some validation using `tf.assert` (we would need to check what the performance implications here are, but I don't think it would be noticeable). Then we can think about using the `kernel_quantizer=int` as a shorthand to that pseudo quantizer much like the string representations of `ste_sign`. We could enable validation by default, but through the `NoOpQuantizer` class advanced users would be able to turn it off and have full flexibility.

**Comment by koenhelwegen** [comment]
I'm not too worried about additional operations - I think the performance degradation will be negligible and would prefer (more) robust behaviour (I realize we cannot cover all cases, but introducing the possibility of a vacuous `kernel_quantizer=1` seems like a bad idea).

> Implicitly quantizing might not be what a users is trying to achieve and is also something that is not clearly visible via the API. E.g. You could think of a scaling "quantizer" lambda x: alpha*x with a trainable variable alpha where x is updated via Bop and alpha via a normal optimizer or a moving average (sorry for the contrived example). Things like that would not be possible to implement since larq would still quantize it to -1, +1.

I'm not sure if I understand you correctly. The user can still set a quantizer and overwrite the default. It is just that in a quantized execution scope, a QuantizedVariable object with precision 1 should never return a bunch of floating points.

> How would you feel about a NoOpQuantizer (probably needs a better name), that is used instead of an integer? It could to some validation using tf.assert (we would need to check what the performance implications here are, but I don't think it would be noticeable). Then we can think about using the kernel_quantizer=int as a shorthand to that pseudo quantizer much like the string representations of ste_sign. We could enable validation by default, but through the NoOpQuantizer class advanced users would be able to turn it off and have full flexibility.

I would rather avoid this; i think `kernel_quantizer=1` is already stretching the meaning of 'quantizer'. If we want to use it just a marker for bop, I would prefer adding a `precision` argument to the layer.

**Comment by lgeiger** [comment]
> I'm not too worried about additional operations - I think the performance degradation will be negligible

I can get some rough performance numbers next week, but unless we have tested this claim, I'd be against adding implicit behaviour there.

@koenhelwegen Since you raise concerns about the overall API design, I am closing this PR for now. We discussed possible options in https://github.com/larq/larq/issues/306#issuecomment-553978646 and agreed on this (or the `NoOpQuantizer` as the best options). Please add your proposed API design there so we can revisit this discussion and go from there.
